lib/frame_window_filtering/hog_labeling.cpp

```diff
@@ -0,0 +1,152 @@
+/// \file
+/// Maintainer: Felice Serena
+///
+///
+
+#include "hog_labeling.h"
+
+#include <opencv2/core/core.hpp>
+#include <opencv2/core/eigen.hpp>
+#include <opencv2/objdetect/objdetect.hpp>
+
+#include <boost/log/trivial.hpp>
+
+#include <classifier/knn.h>
+
+namespace MouseTrack {
+
+/// Returns a n x 2 matrix of (x,y) pairs of valid corner coordinates for
+/// sliding windows of the desired size such that they are within the given
+/// image dimensions.
+///
+std::vector<cv::Point> slidingWindows(int imgWidth, int imgHeight, int stepSize,
+                                      int windowWidth, int windowHeight) {
+  std::vector<cv::Point> result;
+  for (int x = 0; x < imgWidth - windowWidth; x += stepSize) {
+    for (int y = 0; y < imgHeight - windowHeight; y += stepSize) {
+      result.push_back(cv::Point(x, y));
+    }
+  }
+  return result;
+} // namespace MouseTrack
+
+void HogLabeling::train(const Mat &X_train, const Vec &y_train) {
+  auto ptr = std::make_unique<KnnClassifier>();
+  ptr->fit(X_train, y_train);
+  ptr->k(11);
+  _classifier = std::move(ptr);
+  _numLabels = y_train.maxCoeff() + 1;
+}
+
+FrameWindow HogLabeling::operator()(const FrameWindow &window) const {
+  if (window.frames().empty()) {
+    return window;
+  }
+  if (_classifier.get() == nullptr) {
+    BOOST_LOG_TRIVIAL(warning)
+        << "HOG classifier not trained, no labeling performed.";
+    return window;
+  }
+  FrameWindow result = window;
+  // create matrices for labels
+  for (Frame &f : result.frames()) {
+    f.labels.resize(_numLabels);
+    for (auto &l : f.labels) {
+      l.setZero(f.referencePicture.rows(), f.referencePicture.cols());
+    }
+  }
+
+  // hog settings
+  cv::Size windowSize(_windowWidth, _windowHeight);
+  cv::Size blockSize(_blockWidth, _blockHeight);
+  cv::Size blockStride(_blockStrideWidth, _blockStrideHeight);
+  cv::Size cellSize(_cellWidth, _cellHeight);
+
+  // create hog descriptor: it is able to tansform images patches to feature
+  // vectors
+  cv::HOGDescriptor hog(windowSize, blockSize, blockStride, cellSize, _nbins);
```

Comment:
This would definitely clean up as less variables would be floating around, but it would also expose openCV (and hence the implementation) to the interface, which might not be desirable